src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js

```diff
@@ -138,15 +141,36 @@ const clientCleanRender = (element, errorCount = 0) => {
 const clientRenderOnServerString = async (element, errorCount = 0) => {
   const markup = await renderIntoString(element, errorCount);
   resetModules();
+
   var domElement = document.createElement('div');
   domElement.innerHTML = markup;
-  const serverElement = domElement.firstChild;
-  const clientElement = await renderIntoDom(element, domElement, errorCount);
-  // assert that the DOM element hasn't been replaced.
-  // Note that we cannot use expect(serverElement).toBe(clientElement) because
-  // of jest bug #1772
-  expect(serverElement === clientElement).toBe(true);
-  return clientElement;
+  let serverNode = domElement.firstChild;
+
+  const firstClientNode = await renderIntoDom(element, domElement, errorCount);
+  let clientNode = firstClientNode;
+
+  // Make sure all top level nodes match up
+  while (serverNode || clientNode) {
+    expect(serverNode != null).toBe(true);
+    expect(clientNode != null).toBe(true);
+    expect(clientNode.nodeType).toBe(serverNode.nodeType);
+    if (clientNode.nodeType === TEXT_NODE_TYPE) {
+      // Text nodes are stateless so we can just compare values.
+      // This works around a current issue where hydration replaces top-level
+      // text node, but otherwise works.
+      // TODO: we can remove this branch if we add explicit hydration opt-in.
+      // https://github.com/facebook/react/issues/10189
+      expect(serverNode.nodeValue).toBe(clientNode.nodeValue);
```

Comment:
Special case because `shouldReuseContent()` bails out. Happy to change this to use hydration API per conversation with @sebmarkbage, but I'd like to do this as followup.